src/components/pages/news/TableOfContents.tsx

```diff
@@ -0,0 +1,175 @@
+import { useEffect, useState } from 'react'
+import styled from 'styled-components'
+
+import { useTableOfContents } from '../../../hooks/useTableOfContents'
+import { useTocIntersection } from '../../../hooks/useTocIntersection'
+
+interface TableOfContentsProps {
+  htmlContent: string
+}
+
+const TableOfContents = ({ htmlContent }: TableOfContentsProps) => {
+  const tocItems = useTableOfContents(htmlContent)
+  const { activeId, highlightedId, handleClick } = useTocIntersection(tocItems)
+  const [isScrolledDown, setIsScrolledDown] = useState<boolean>(false)
+
+  useEffect(() => {
+    if (tocItems.length === 0) return
+
+    const articleElement = document.querySelector('article')
+    if (!articleElement) return
+
+    articleElement.style.scrollPaddingTop = '120px'
+    document.documentElement.style.scrollPaddingTop = '120px'
+    document.body.style.scrollPaddingTop = '120px'
+
+    const headers = articleElement.querySelectorAll('h2, h3')
+    headers.forEach((header, index) => {
+      const tocItem = tocItems[index]
+      if (tocItem && !header.id) {
+        header.id = tocItem.id
+        header.setAttribute('data-toc-header', 'true')
+      }
+    })
+
+    return () => {
+      if (articleElement) {
+        articleElement.style.scrollPaddingTop = ''
+      }
+      document.documentElement.style.scrollPaddingTop = ''
+      document.body.style.scrollPaddingTop = ''
+    }
+  }, [tocItems])
+
+  useEffect(() => {
+    const tocContainer = document.querySelector('[data-toc-container]') as HTMLElement
+    if (!tocContainer) return
+
+    const handleTocScroll = () => {
+      const scrollTop = tocContainer.scrollTop
+      setIsScrolledDown(scrollTop > 10)
+    }
+
+    tocContainer.addEventListener('scroll', handleTocScroll, { passive: true })
+    handleTocScroll()
+
+    return () => {
+      tocContainer.removeEventListener('scroll', handleTocScroll)
+    }
+  }, [tocItems])
+
+  useEffect(() => {
+    if (!highlightedId) return
+
+    const targetHeader = document.getElementById(highlightedId)
+    if (targetHeader) {
+      targetHeader.style.color = '#fbbf24'
+    }
+  }, [highlightedId])
+
+  useEffect(
+    () => () => {
+      if (highlightedId) {
+        const targetHeader = document.getElementById(highlightedId)
+        if (targetHeader) {
+          targetHeader.style.color = ''
+        }
+      }
+    },
+    [highlightedId]
+  )
```

Comment:
This is a bit confusing to me. The effect is returning a function, so I guess this is a cleanup function??